File: xr-plugin/com.holoi.xr.holokit/Runtime/HoloKit/HoloKitProfile.cs
using UnityEngine;

namespace HoloKit
{
    // TODO: Get HoloKitType from NFC?
    public enum HoloKitType
    {
        HoloKitX = 0
    }

    public enum PhoneType
    {
        iPhoneXS = 0,
        iPhoneXSMax = 1,
        iPhone11Pro = 2,
        iPhone11ProMax = 3,
        iPhone12 = 4,
        iPhone12Pro = 5,
        iPhone12ProMax = 6,
        iPhone13 = 7,
        iPhone13Pro = 8,
        iPhone13ProMax = 9,
        iPad = 10, // all iPads are the same to us
        Unknown = 11
    }

    public struct HoloKitModel
    {
        /// <summary>
        /// Distance beetween eyes
        /// </summary>
        public float OpticalAxisDistance;

        /// <summary>
        /// 3D offset from the center of bottomline of the holokit phone display to the center of two eyes.
        /// </summary>
        public Vector3 MrOffset;

        /// <summary>
        /// Eye view area width
        /// </summary>
        public float ViewportInner;

        /// <summary>
        /// Eye view area height
        /// </summary>
        public float ViewportOuter;

        /// <summary>
        /// Eye view area spillter width
        /// </summary>
        public float ViewportTop;

        /// <summary>
        /// Eye view area spillter width
        /// </summary>
        public float ViewportBottom;

        /// <summary>
        /// Fresnel lens focal length
        /// </summary>
        public float FocalLength;

        /// <summary>
        /// Screen To Fresnel distance
        /// </summary>
        public float ScreenToLens;

        /// <summary>
        /// Fresnel To eye distance
        /// </summary>
        public float LensToEye;

        /// <summary>
        /// Bottom of the holder to bottom of the view
        /// </summary>
        public float AxisToBottom;

        /// <summary>
        /// The distance between the center of the HME and the marker
        /// </summary>
        public float HorizontalAlignmentMarkerOffset;
    }

    public struct PhoneModel
    {
        /// <summary>
        /// The long screen edge of the phone
        /// </summary>
        public float ScreenWidth;

        /// <summary>
        /// The short screen edge of the phone
        /// </summary>
        public float ScreenHeight;

        /// <summary>
        /// The distance from the bottom of display area to the touching surface of the holokit phone holder
        /// </summary>
        public float ScreenBottom;

        /// <summary>
        /// The distance from the center of the display to the rendering center
        /// </summary>
        public float CenterLineOffset;

        /// <summary>
        /// The 3D offset vector from center of the camera to the center of the display area's bottomline
        /// </summary>
        public Vector3 CameraOffset;
    }
    public static class HoloKitProfile
    {
        public static HoloKitModel GetHoloKitModel(HoloKitType holokitType)
        {
            switch(holokitType)
            {
                case HoloKitType.HoloKitX:
                    return new HoloKitModel
                    {
                        OpticalAxisDistance = 0.064f,
                        MrOffset = new Vector3(0f, -0.02894f, -0.07055f),
                        ViewportInner = 0.0292f,
                        ViewportOuter = 0.0292f,
                        ViewportTop = 0.02386f,
                        ViewportBottom = 0.02386f,
                        FocalLength = 0.065f,
                        ScreenToLens = 0.02715f + 0.03136f + 0.002f,
                        LensToEye = 0.02497f + 0.03898f,
                        AxisToBottom = 0.02990f,
                        HorizontalAlignmentMarkerOffset = 0.05075f,
                    };
                default:
                    return new HoloKitModel();
            }
        }

        public static PhoneModel GetPhoneModel()
        {
            PhoneType phoneType = GetPhoneType();
            switch (phoneType)
            {
                case PhoneType.iPhoneXS:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.135097f,
                        ScreenHeight = 0.062391f,
                        ScreenBottom = 0.00391f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.05986f, -0.055215f, -0.0091f)
                    };
                case PhoneType.iPhoneXSMax:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.14971f,
                        ScreenHeight = 0.06961f,
                        ScreenBottom = 0.00391f,
                        CenterLineOffset = -0.006f,
                        CameraOffset = new Vector3(0.06694f, -0.09405f, -0.00591f)
                    };
                case PhoneType.iPhone11Pro:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.13495f,
                        ScreenHeight = 0.06233f,
                        ScreenBottom = 0.00452f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.05996f, -0.02364f - 0.03494f, -0.00591f)
                    };
                case PhoneType.iPhone11ProMax:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.14891f,
                        ScreenHeight = 0.06881f,
                        ScreenBottom = 0.00452f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.066945f, -0.061695f, -0.0091f)
                    };
                case PhoneType.iPhone12:
                case PhoneType.iPhone13:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.13977f,
                        ScreenHeight = 0.06458f,
                        ScreenBottom = 0.00347f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.05996f, -0.02364f - 0.03494f, -0.00591f)
                    };
                case PhoneType.iPhone12Pro:
                case PhoneType.iPhone13Pro:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.13977f,
                        ScreenHeight = 0.06458f,
                        ScreenBottom = 0.00347f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.05996f, -0.02364f - 0.03494f, -0.00591f)
                    };
                case PhoneType.iPhone12ProMax:
                case PhoneType.iPhone13ProMax:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.15390f,
                        ScreenHeight = 0.07113f,
                        ScreenBottom = 0.00347f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.066945f, -0.061695f, -0.0091f)
                    };
                case PhoneType.iPad:
                case PhoneType.Unknown:
                default:
                    return new PhoneModel
                    {
                        ScreenWidth = 0.15390f,
                        ScreenHeight = 0.07113f,
                        ScreenBottom = 0.00347f,
                        CenterLineOffset = 0f,
                        CameraOffset = new Vector3(0.066945f, -0.061695f, -0.0091f)
                    };
            }
        }

        public static PhoneType GetPhoneType()
        {
            return SystemInfo.deviceModel switch
            {
                // iPhones
                "iPhone11,2" => PhoneType.iPhoneXS,
                "iPhone11,4" => PhoneType.iPhoneXSMax,
                "iPhone11,6" => PhoneType.iPhoneXSMax,
                "iPhone12,3" => PhoneType.iPhone11Pro,
                "iPhone12,5" => PhoneType.iPhone11ProMax,
                "iPhone13,2" => PhoneType.iPhone12,
                "iPhone13,3" => PhoneType.iPhone12Pro,
                "iPhone13,4" => PhoneType.iPhone12ProMax,
                "iPhone14,5" => PhoneType.iPhone13,
                "iPhone14,2" => PhoneType.iPhone13Pro,
                "iPhone14,3" => PhoneType.iPhone13ProMax,
                // iPads
                // TODO: Add more iPads
                "iPad13,8" => PhoneType.iPad,
                "iPad13,9" => PhoneType.iPad,
                "iPad13,10" => PhoneType.iPad,
                "iPad13,11" => PhoneType.iPad,
                // Not supported devices
                _ => PhoneType.Unknown
            };
        }
    }
}

File: HoloKitSDK/HoloKitProfile.m
//
//  HoloKitProfile.m
//  holokit-sdk
//
//  Created by Yuchen Zhang on 2022/7/18.
//

#import "HoloKitProfile.h"
#import <sys/utsname.h>

@interface HoloKitProfile()

@end

@implementation HoloKitProfile

+ (HoloKitModel)getHoloKitModel:(HoloKitType)holokitType {
    HoloKitModel holokitModel;
    switch (holokitType) {
        case HoloKitX:
            holokitModel.OpticalAxisDistance = 0.064;
            holokitModel.MrOffset = simd_make_float3(0, -0.02894, -0.07055);
            holokitModel.ViewportInner = 0.0292;
            holokitModel.ViewportOuter = 0.0292;
            holokitModel.ViewportTop = 0.02386;
            holokitModel.ViewportBottom = 0.02386;
            holokitModel.FocalLength = 0.065;
            holokitModel.ScreenToLens = 0.02715 + 0.03136 + 0.002;
            holokitModel.LensToEye = 0.02497 + 0.03898;
            holokitModel.AxisToBottom = 0.02990;
            holokitModel.HorizontalAlignmentMarkerOffset = 0.05075;
            break;
        default:
            break;
    }
    return holokitModel;
}

+ (PhoneType)getPhoneType {
    struct utsname systemInfo;
    uname(&systemInfo);
    NSString *deviceName = [NSString stringWithCString:systemInfo.machine encoding:NSUTF8StringEncoding];
    if ([deviceName isEqualToString:@"iPhone11,2"]) {
        return iPhoneXS;
    } else if ([deviceName isEqualToString:@"iPhone11,4"] || [deviceName isEqualToString:@"iPhone11,6"]) {
        return iPhoneXSMax;
    } else if ([deviceName isEqualToString:@"iPhone12,3"]) {
        return iPhone11Pro;
    } else if ([deviceName isEqualToString:@"iPhone12,5"]) {
        return iPhone11ProMax;
    } else if ([deviceName isEqualToString:@"iPhone13,1"]) {
        return iPhone12mini;
    } else if ([deviceName isEqualToString:@"iPhone13,2"]) {
        return iPhone12;
    } else if ([deviceName isEqualToString:@"iPhone13,3"]) {
        return iPhone12Pro;
    } else if ([deviceName isEqualToString:@"iPhone13,4"]) {
        return iPhone12ProMax;
    } else if ([deviceName isEqualToString:@"iPhone14,4"]) {
        return iPhone13mini;
    } else if ([deviceName isEqualToString:@"iPhone14,5"]) {
        return iPhone13;
    } else if ([deviceName isEqualToString:@"iPhone14,2"]) {
        return iPhone13Pro;
    } else if ([deviceName isEqualToString:@"iPhone14,3"]) {
        return iPhone13ProMax;
    } else if ([deviceName isEqualToString:@"iPhone14,7"]) {
        return iPhone14;
    } else if ([deviceName isEqualToString:@"iPhone14,8"]) {
        return iPhone14Plus;
    } else if ([deviceName isEqualToString:@"iPhone15,2"]) {
        return iPhone14Pro;
    } else if ([deviceName isEqualToString:@"iPhone15,3"]) {
        return iPhone14ProMax;
    } else if ([deviceName isEqualToString:@"iPad13,8"] ||
               [deviceName isEqualToString:@"iPad13,9"] ||
               [deviceName isEqualToString:@"iPad13,10"] ||
               [deviceName isEqualToString:@"iPad13,11"]) {
        return iPad;
    } else {
        return Unknown;
    }
}

+ (PhoneModel)getPhoneModel {
    PhoneModel phoneModel;
    switch ([HoloKitProfile getPhoneType])
    {
        case iPhoneXS:
            phoneModel.ScreenWidth = 0.135097;
            phoneModel.ScreenHeight = 0.062391;
            phoneModel.ScreenBottom = 0.00391;
            phoneModel.CameraOffset = simd_make_float3(0.05986, -0.055215, -0.0091);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhoneXSMax:
            phoneModel.ScreenWidth = 0.14971;
            phoneModel.ScreenHeight = 0.06961;
            phoneModel.ScreenBottom = 0.00391;
            phoneModel.CameraOffset = simd_make_float3(0.06694, -0.09405, -0.00591);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone11Pro:
            phoneModel.ScreenWidth = 0.13495;
            phoneModel.ScreenHeight = 0.06233;
            phoneModel.ScreenBottom = 0.00452;
            phoneModel.CameraOffset = simd_make_float3(0.059955, -0.05932, -0.00591);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone11ProMax:
            phoneModel.ScreenWidth = 0.14891;
            phoneModel.ScreenHeight = 0.06881;
            phoneModel.ScreenBottom = 0.00452;
            phoneModel.CameraOffset = simd_make_float3(0.066935, -0.0658, -0.00591);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone12mini:
            phoneModel.ScreenWidth = 0.12496;
            phoneModel.ScreenHeight = 0.05767;
            phoneModel.ScreenBottom = 0.00327;
            phoneModel.CameraOffset = simd_make_float3(0.05508, -0.05354, -0.00620);
            phoneModel.ScreenDpi = 476;
            break;
        case iPhone12:
            phoneModel.ScreenWidth = 0.13977;
            phoneModel.ScreenHeight = 0.06458;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.060625, -0.05879, -0.00633);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone12Pro:
            phoneModel.ScreenWidth = 0.13977;
            phoneModel.ScreenHeight = 0.06458;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.061195, -0.05936, -0.00551);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone12ProMax:
            phoneModel.ScreenWidth = 0.15390;
            phoneModel.ScreenHeight = 0.07113;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.04952, -0.06464, -0.00591);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone13mini:
            phoneModel.ScreenWidth = 0.12496;
            phoneModel.ScreenHeight = 0.05767;
            phoneModel.ScreenBottom = 0.00327;
            phoneModel.CameraOffset = simd_make_float3(0.0549, -0.05336, -0.00633);
            phoneModel.ScreenDpi = 476;
            break;
        case iPhone13:
            phoneModel.ScreenWidth = 0.13977;
            phoneModel.ScreenHeight = 0.06458;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.06147, -0.05964, -0.00781);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone13Pro:
            phoneModel.ScreenWidth = 0.13977;
            phoneModel.ScreenHeight = 0.06458;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.042005, -0.05809, -0.00727);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone13ProMax:
            phoneModel.ScreenWidth = 0.15390;
            phoneModel.ScreenHeight = 0.07113;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.04907, -0.06464, -0.00727);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone14:
            phoneModel.ScreenWidth = 0.13977;
            phoneModel.ScreenHeight = 0.06458;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.061475, -0.05964, -0.00848);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone14Plus:
            phoneModel.ScreenWidth = 0.15390;
            phoneModel.ScreenHeight = 0.07113;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.06787, -0.06552, -0.00851);
            phoneModel.ScreenDpi = 458;
            break;
        case iPhone14Pro:
            phoneModel.ScreenWidth = 0.14109;
            phoneModel.ScreenHeight = 0.06508;
            phoneModel.ScreenBottom = 0.003185;
            phoneModel.CameraOffset = simd_make_float3(0.04021, -0.05717, -0.00784);
            phoneModel.ScreenDpi = 460;
            break;
        case iPhone14ProMax:
            phoneModel.ScreenWidth = 0.15434;
            phoneModel.ScreenHeight = 0.07121;
            phoneModel.ScreenBottom = 0.003185;
            phoneModel.CameraOffset = simd_make_float3(0.046835, -0.0633, -0.0078);
            phoneModel.ScreenDpi = 460;
            break;
        case iPad:
        case Unknown:
        default:
            phoneModel.ScreenWidth = 0.15390;
            phoneModel.ScreenHeight = 0.07113;
            phoneModel.ScreenBottom = 0.00347;
            phoneModel.CameraOffset = simd_make_float3(0.066945, -0.061695, -0.0091);
            phoneModel.ScreenDpi = 460;
            break;
    }
    return phoneModel;
}

+ (BOOL)IsCurrentDeviceSupportedByHoloKit {
    PhoneType phoneType = [HoloKitProfile getPhoneType];
    return phoneType != Unknown;
}

+ (BOOL)IsCurrentDeviceIpad {
    PhoneType phoneType = [HoloKitProfile getPhoneType];
    return phoneType == iPad;
}

+ (BOOL)IsCurrentDeviceEquippedWithLiDAR {
    PhoneType phoneType = [HoloKitProfile getPhoneType];
    if (phoneType == iPhone12Pro) {
        return true;
    }
    if (phoneType == iPhone12ProMax) {
        return true;
    }
    if (phoneType == iPhone13Pro) {
        return true;
    }
    if (phoneType == iPhone13ProMax) {
        return true;
    }
    if (phoneType == iPhone14Pro) {
        return true;
    }
    if (phoneType == iPhone14ProMax) {
        return true;
    }
    return false;
}

@end
